File: pixivSearch.py
#!/usr/bin/python
# coding:utf-8

import requests
import threading
import sys
import math
import Queue
import re

class geturl(threading.Thread):
    def __init__(self,name,keyword,number):
        threading.Thread.__init__(self)
        self.name=name
        self.keyword=keyword
        self.number=number
    def run(self):
        global workqueue
        global urls
        while not workqueue.empty():
            lock.acquire()
            item=workqueue.get()
            lock.release()
            try:
                html = s.get('http://www.pixiv.net/search.php?word='+self.keyword+'&p='+str(item))
                lis = re.findall('<a href="/bookmark_detail.php\?illust_id=(.*?)" class="bookmark-count _ui-tooltip" data-tooltip=.*?(\w*)</a>', html.text)
                for li in lis:
                    if int(li[1]) >=self.number:
                        lock.acquire()
                        urls.append('http://www.pixiv.net/member_illust.php?mode=medium&illust_id='+li[0])
                        lock.release()
            except BaseException,e:
                print 'page'+str(item)+'出错'
                lock.acquire()
                item=workqueue.put(item)
                lock.release()
#标签
keyword = sys.argv[1]
#收藏数
number = int(sys.argv[2])
#登录pixiv
s = requests.Session()
payload = {'mode':'login','return_to':'/','pixiv_id':'','pass':'','skip':'1'}
s.post('https://www.pixiv.net/login.php',data=payload)

print '标签为 '+keyword+' 收藏数大于等于 '+str(number)+' 的链接'
html = s.get('http://www.pixiv.net/search.php?word='+keyword)
pages = re.findall('<span class="count-badge">(\w*).*</span>', html.text)[0]
pages = int(math.ceil(float(pages)/20))
print '总页数为 '+str(pages)

lock = threading.Lock()
workqueue = Queue.Queue()
threadnum = 100
threads = list()
urls = list()

for task in range(1,pages+1):
    workqueue.put(task)

for i in range(threadnum):
    thread_name = 'thread%s' %i
    thread = geturl(thread_name,keyword,number)
    thread.start()
    threads.append(thread)
    
for i in threads:
    i.join()

for i in urls:
    print i
print '完成'+str(len(urls))

File: pixivSearch.js
#!/usr/bin/node

var request=require('request');
var keyword=process.argv[2];
var number=process.argv[3];

function login(){
    console.log('标签为 '+keyword+' 收藏数大于等于 '+number+' 的链接');
    return new Promise(function(resolve,reject){
        request=request.defaults({jar:true});
        request({
            url:'https://www.pixiv.net/login.php',
            method:'POST',
            form:{
                mode:'login',
                return_to:'/',
                pixiv_id:'',
                pass:'',
                skip:'1',
            },
            followAllRedirects:true,
        },function(err,res,body){
            resolve(request);
        });
    });
}

function getPages(request){
    return new Promise(function(resolve,reject){
        request({
            url:encodeURI('http://www.pixiv.net/search.php?word='+keyword),
            method:"GET",
        },function(err,res,body){
            require("jsdom").env(body, function(err, window) {
                var $ = require("jquery")(window);
                var pages = $('span.count-badge').text();
                pages = Math.ceil((+pages.slice(0,-1))/20);
                console.log('总页数为 '+pages);
                request.pages=pages;
                resolve(request);
            });
        });
    });
}

function getUrl(request){
    for(var p=1;p<=request.pages;p++){
        request({
            method:'GET',
            url:encodeURI('http://www.pixiv.net/search.php?word='+keyword+'&p='+p),
        },function(err,res,body){
            require("jsdom").env(body, function(err, window) {
                var $ = require("jquery")(window);
                var li = $('li.image-item');
                for(var l=0;l<li.length;l++){
                    var bookmark_a=$(li[l]).find('a.bookmark-count._ui-tooltip');
                    var num = +$(bookmark_a).text();
                    if(num>=number){
                        console.log('http://www.pixiv.net'+$($(li[l]).children('a')[0]).attr('href'));
                    }
                }
            });
        });
    }
}

login().then(getPages).then(getUrl);